### PR TITLE
feat(RevenueCatUI): arbitrate web_view drag gestures with the paywall scroll

### PR DESCRIPTION
**Pull request by AlvaroBrey**
### Checklist
- [x] If applicable, unit tests
- [ ] If applicable, create follow-up issues for `purchases-ios` and hybrids

### Motivation
A `web_view` embedded in the paywall double-scrolls: dragging interactive or scrollable content (a map, an inner list) also scrolls the whole paywall.

### Description
`PaywallWebView` decides, once per drag, whether the web_view or the paywall scroll owns the gesture, via `requestDisallowInterceptTouchEvent` (claim-once, since Compose won't hand a gesture back). The web_view claims only when its content consumes the drag:
- a per-gesture JS probe reports inner `overflow` scrollers and `touch-action` maps — the signals native code can't see — over the existing secure message channel;
- native `canScrollVertically`/`Horizontally` covers root/page scroll;
- otherwise the paywall scrolls.

No-ops to native scrollability on old WebViews without the probe features. Verified on device (inner list, carousel, map) plus unit tests for the decision logic.

---

> [!NOTE]
> **Medium Risk**
> Touch handling and injected JS affect all paywall web_views; behavior degrades to native scroll-only on older WebViews, with edge cases where JS-only gesture blocking isn't detected.
> 
> **Overview**
> Fixes **double-scroll** when dragging inside paywall `web_view` content (maps, inner lists, carousels) by deciding once per drag whether the WebView or the outer paywall scroll owns the gesture.
> 
> Introduces **`PaywallWebView`**, which calls `requestDisallowInterceptTouchEvent` only when **`shouldWebViewOwnGesture`** says the WebView should own the drag (respecting touch slop and dominant axis). A **`WebViewGestureOwnershipProbe`** injects document-start JS and uses `WebMessageListener` to report per-touch whether inner scrollers or non-default `touch-action` need the gesture; that verdict overrides native checks when it says "own," otherwise root **`canScrollVertically`/`Horizontally`** decides.
> 
> **`WebViewComponentView`** swaps plain `WebView` for `PaywallWebView`, wires install/teardown of the probe on release, and **`disableTapHighlight`** no longer falls back to `*` when origin is unknown—it skips injection instead (tests updated). Unit tests cover arbitration rules and probe install behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ab663d1da4f82dda3fdc9c3ea852f9a75d49586c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by AlvaroBrey** [comment]
@cursor review

**Comment by AlvaroBrey** [comment]
@cursor review

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+          var s = getComputedStyle(n);
+          if (s.touchAction && s.touchAction !== 'auto' && s.touchAction !== 'manipulation') return true;
+          if ((s.overflowY === 'auto' || s.overflowY === 'scroll') && n.scrollHeight > n.clientHeight) return true;
+          if ((s.overflowX === 'auto' || s.overflowX === 'scroll') && n.scrollWidth > n.clientWidth) return true;
+        }
+        return false;
+      }
+      document.addEventListener('touchstart', function (event) {
+        try {
+          window[name].postMessage(consumesGesture(event.target) ? '$VERDICT_OWN' : 'release');
```
[the remark this answers is not in the document]
Comment (AlvaroBrey) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt:32]: Intentional, not changing. The probe fires on `touchstart`, before any drag direction exists, so it cannot be direction-aware at claim time. Compose does not hand a gesture back mid-drag ([issuetracker 174348612](https://issuetracker.google.com/issues/174348612)), which is why arbitration is claim-once with no edge-chaining. Deferring to the paywall would lose inner scrolling for the whole gesture; claiming means we cannot release at the inner edge. We deliberately favor the touched inner scroller. This is the accepted tradeoff, device-verified.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/PaywallWebView.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+@Suppress("ReturnCount", "LongParameterList")
+internal fun shouldWebViewOwnGesture(
+    totalDx: Float,
+    totalDy: Float,
+    touchSlop: Int,
+    webContentWantsGesture: Boolean?,
+    canScrollHorizontally: (direction: Int) -> Boolean,
+    canScrollVertically: (direction: Int) -> Boolean,
+): Boolean {
+    if (webContentWantsGesture == true) return true
```
[the remark this answers is not in the document]
Comment (AlvaroBrey) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/PaywallWebView.kt:28]: Intentional, not changing. The probe fires on `touchstart`, before any drag direction exists, so it cannot be direction-aware at claim time. Compose does not hand a gesture back mid-drag ([issuetracker 174348612](https://issuetracker.google.com/issues/174348612)), which is why arbitration is claim-once with no edge-chaining. Deferring to the paywall would lose inner scrolling for the whole gesture; claiming means we cannot release at the inner edge. We deliberately favor the touched inner scroller. This is the accepted tradeoff, device-verified.

**Comment by AlvaroBrey** [comment]
* **#3787** <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3787?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3824** <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3824?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3823** <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3823?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3823?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ * purely via a non-passive `preventDefault` without a `touch-action` isn't detected.
+ */
+@Suppress("TooGenericExceptionCaught")
+internal fun WebView.installGestureOwnershipProbe(
+    expectedOrigin: String?,
+    onVerdict: (wantsGesture: Boolean) -> Unit,
+) {
+    if (!WebViewFeature.isFeatureSupported(WebViewFeature.WEB_MESSAGE_LISTENER)) return
+    if (!WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) return
+    val allowedOrigins = setOf(expectedOrigin ?: "*")
```
Comment (ajpallares) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt]: Would it be better to just `return` if `expectedOrigin` is `null`? Maybe not a problem, but in theory, `expectedOrigin` should never be `null` at this point, right?

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+// On each touchstart, reports "own" if the touched element or an ancestor scrolls or declares a
+// non-default `touch-action` — the per-element signals native canScroll can't see. Passive.
+private val GESTURE_PROBE_SCRIPT =
+    """
+    (function () {
+      var name = '$GESTURE_PROBE_OBJECT_NAME';
+      function consumesGesture(el) {
+        var node = el && el.nodeType === 1 ? el : (el ? el.parentElement : null);
+        for (var n = node; n && n.nodeType === 1; n = n.parentElement) {
```
Comment (ajpallares) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt]: Probably obvious for anyone with knowledge, but maybe this reads better for those of us that aren't experts 😅 

```suggestion
        var ELEMENT_NODE = Node.ELEMENT_NODE;
        var node = el && el.nodeType === ELEMENT_NODE ? el : (el ? el.parentElement : null);
        for (var n = node; n && n.nodeType === ELEMENT_NODE; n = n.parentElement) {
```

**Review by ajpallares — APPROVED** [review]
Nice one! Two small comments, non-blocking at all

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt (hunk 3 above)
Comment (AlvaroBrey) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt]: Yeah, it's cleaner to just return, thanks. Done

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt (hunk 4 above)
Comment (AlvaroBrey) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/webview/WebViewGestureOwnershipProbe.kt]: Done!
